lib/traiderdaive/traiderdaive/gym_environments/full_hyperdrive_env.py

```diff
@@ -0,0 +1,490 @@
+                    # TODO use logging here
+                    print(f"Warning: Failed to open trade: {err=}")
+                    # Terminate if error
+                    return True
+
+        # LP actions
+        lp_actions_expit = expit(action[-4:])
+        add_lp_probability = lp_actions_expit[0]
+        add_lp_volume = FixedPoint(lp_actions_expit[1]) * self.gym_config.max_trade_amount
+        remove_lp_probability = lp_actions_expit[2]
+        remove_lp_volume = FixedPoint(lp_actions_expit[3]) * self.gym_config.max_trade_amount
+
+        try:
+            if add_lp_probability > self.gym_config.open_threshold:
+                self.rl_bot.add_liquidity(add_lp_volume)
+            if (
+                remove_lp_probability > self.gym_config.close_threshold
+                and remove_lp_volume <= self.rl_bot.wallet.lp_tokens
+            ):
+                self.rl_bot.remove_liquidity(remove_lp_volume)
+            # Always try and remove withdrawal shares
+            if self.rl_bot.wallet.withdraw_shares > 0:
+                # TODO error handling or check when withdrawal shares are not withdrawable
+                self.rl_bot.redeem_withdraw_share(self.rl_bot.wallet.withdraw_shares)
+        except Exception as err:  # pylint: disable=broad-except
+            # TODO use logging here
+            print(f"Warning: Failed to LP: {err=}")
+            # Terminate if error
+            return True
+
+        return False
+
+    def step(self, action: np.ndarray) -> tuple[dict[str, np.ndarray], float, bool, bool, dict[str, Any]]:
+        """Takes a step in the the environment.
+
+        Arguments
+        ---------
+        action: ActType
+            An action provided by the agent to update the environment state
+
+        Returns
+        -------
+        tuple[np.ndarray, float, bool, bool, dict[str, Any]]
+            Contains the following
+
+            observation: ObsType
+                An element of the environment's observation_space.
+            reward: float
+                Reward for taking the action.
+            terminated: bool
+                Whether the agent reaches the terminal state, which can be positive or negative.
+                If true, user needs to call reset
+            truncated: bool
+                Whether the truncation condition outside the scope of the MDP is satisfied,
+                e.g., timelimit, or agent going out of bounds.
+                If true, user needs to call reset
+            info: dict[str, Any]
+                Contains auxiliary diagnostic information for debugging, learning, logging.
+        """
+
+        truncated = self._apply_action(action)
+
+        # Run other bots
+        for random_bot in self.random_bots:
+            try:
+                random_bot.execute_policy_action()
+            except Exception as err:  # pylint: disable=broad-except
+                print(f"Warning: Failed to execute random bot: {err=}")
```

Comment:
this should be logging.warning right?

Reply:
probably, I didn't want to set up logging in this file and just wanted to print a warning when this happens

Reply:
fine 🙃